libraries/psibase/tester/src/checkSchema.cpp

```diff
@@ -0,0 +1,157 @@
+#include <psibase/checkSchema.hpp>
+
+#include <algorithm>
+#include <format>
+#include <psibase/package.hpp>
+#include <psibase/tester.hpp>
+#include <ranges>
+#include <unordered_map>
+
+namespace psibase
+{
+   std::optional<std::string> findDifference(const ServiceSchema& lhs, const ServiceSchema& rhs)
+   {
+      auto matcher = psio::schema_types::TypeMatcher{
+          lhs.types,
+          rhs.types,
+          psio::SchemaDifference::equivalent,
+      };
+
+      for (const auto& [name, lty] : lhs.actions)
+      {
+         if (auto rpos = rhs.actions.find(name); rpos != rhs.actions.end())
+         {
+            const auto& rty = rpos->second;
+            if (!matcher.match(lty.params, rty.params))
+               return std::format("Parameter types for action {} do not match", name);
+            if (lty.result && rty.result)
+            {
+               if (!matcher.match(*lty.result, *rty.result))
+                  return std::format("Return type for {} does not match", name);
+            }
+            else if (!lty.result && rty.result)
```

Comment:
Why is this an empty body?
Also this condition `(!lty.result && rty.result)` is repeated